simde/arm/neon/cmla.h

```diff
@@ -0,0 +1,371 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * Copyright:
+ *   2021      Atharva Nimbalkar <atharvakn@gmail.com>
+ */
+
+ #if !defined(SIMDE_ARM_NEON_CMLA_H)
+ #define SIMDE_ARM_NEON_CMLA_H
+
+ #include "types.h"
+
+ HEDLEY_DIAGNOSTIC_PUSH
+ SIMDE_DISABLE_UNWANTED_DIAGNOSTICS
+ SIMDE_BEGIN_DECLS_
+
+ /* * rot 0 * */
+
+ SIMDE_FUNCTION_ATTRIBUTES
+ simde_float32x2_t
+ simde_vcmla_f32(simde_float32x2_t r, simde_float32x2_t a, simde_float32x2_t b) {
+   #if defined(SIMDE_ARM_NEON_A32V8_NATIVE)
+    return vcmla_f32(r, a, b);
+   #else
+    simde_float32x2_private
+      r_ = simde_float32x2_to_private(r),
+      a_ = simde_float32x2_to_private(a),
+      b_ = simde_float32x2_to_private(b);
+
+    SIMDE_VECTORIZE
+    for(size_t i = 0 ; i < (sizeof(r_.values) / sizeof(r_.values[0])) ; i++) {
+      if(i % 2 == 0)
+        r_.values[i] += b_.values[i] * a_.values[i];
+      else
+        r_.values[i] += b_.values[i] * a_.values[i-1];
+    }
+
+    return simde_float32x2_from_private(r_);
+   #endif
+ }
+ #if defined(SIMDE_ARM_NEON_A32V8_ENABLE_NATIVE_ALIASES)
+   #undef vcmla_f32
+   #define vcmla_f32(r, a, b) simde_vcmla_f32(r, a, b)
+ #endif
+
+ SIMDE_FUNCTION_ATTRIBUTES
+ simde_float32x4_t
+ simde_vcmlaq_f32(simde_float32x4_t r, simde_float32x4_t a, simde_float32x4_t b) {
+   #if defined(SIMDE_ARM_NEON_A32V8_NATIVE)
+    return vcmlaq_f32(r, a, b);
+   #else
+    simde_float32x4_private
+      r_ = simde_float32x4_to_private(r),
+      a_ = simde_float32x4_to_private(a),
+      b_ = simde_float32x4_to_private(b);
+
+    SIMDE_VECTORIZE
+    for(size_t i = 0 ; i < (sizeof(r_.values) / sizeof(r_.values[0])) ; i++) {
+      if(i % 2 == 0)
```

Comment:
First, you should prefer `& 1` to `% 2`.  Surprisingly, the doesn't *always* do the right thing even for this simple optimization.

In SIMD programming, it's very important to avoid branches.  Remember, SIMD is "single instruction multiple data"; if you have a branch you're running different instructions for different data.  Often this means that the compiler can't automatically vectorize the code.

In this case, you could use something like this for the body of the loop:

```c
r_.values[i] += b_.values[i] * a_.values[i & 2];
```

Reply:
Ah I see. I think I can remove branching code from all 4 variations of the function. Thanks for pointing this out!